File: icore/src/main/java/com/ddu/icore/rx/activityresult/RxActivityResultFragment.java
package com.ddu.icore.rx.activityresult;


import android.app.Fragment;
import android.content.Intent;
import android.os.Bundle;

import java.util.HashMap;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.PublishSubject;

/**
 * Created by yzbzz on 2018/1/5.
 */

public class RxActivityResultFragment extends Fragment {

    private Map<Integer, PublishSubject<ActivityResultInfo>> mSubjects = new HashMap<>();

    public RxActivityResultFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        PublishSubject<ActivityResultInfo> publishSubject = mSubjects.remove(requestCode);
        if (null != publishSubject) {
            ActivityResultInfo activityResultInfo = new ActivityResultInfo();
            activityResultInfo.setRequestCode(requestCode);
            activityResultInfo.setResultCode(resultCode);
            activityResultInfo.setData(data);
            publishSubject.onNext(activityResultInfo);
            publishSubject.onComplete();
        }
    }

    public Observable<ActivityResultInfo> startForResult(final Intent intent, final int requestCode) {
        PublishSubject<ActivityResultInfo> publishSubject = PublishSubject.create();
        mSubjects.put(requestCode, publishSubject);
        return publishSubject.doOnSubscribe(new Consumer<Disposable>() {
            @Override
            public void accept(Disposable disposable) throws Exception {
                startActivityForResult(intent, requestCode);
            }
        });

    }
}

File: icore/src/main/java/com/ddu/icore/rx/activityresult/RxActivityResultFragment.kt
package com.ddu.icore.rx.activityresult


import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import java.util.*

/**
 * Created by yzbzz on 2018/1/5.
 */

class RxActivityResultFragment : androidx.fragment.app.Fragment() {

    @SuppressLint("UseSparseArrays")
    private val mSubjects = HashMap<Int, PublishSubject<ActivityResultInfo>>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        val publishSubject = mSubjects.remove(requestCode)
        publishSubject?.run {
            val activityResultInfo = ActivityResultInfo()
            activityResultInfo.requestCode = requestCode
            activityResultInfo.resultCode = resultCode
            activityResultInfo.data = data
            onNext(activityResultInfo)
            onComplete()
        }
    }

    fun startForResult(intent: Intent, requestCode: Int): Observable<ActivityResultInfo> {
        val publishSubject = PublishSubject.create<ActivityResultInfo>()
        mSubjects[requestCode] = publishSubject
        return publishSubject.doOnSubscribe { startActivityForResult(intent, requestCode) }

    }
}
